### ISSUE TITLE
[TF 2.0] tf.summary centralized step management

### ISSUE THREAD

**Issue by nfelt** [tensorflow/tensorflow#26406]
This feature request tracks providing a way to centrally control the notion of the "step" of a given summary (used e.g. as the x-axis in TensorBoard scalar plots). In TF 1.x there was the global step, but in TF 2.0 it no longer exists and global state is generally discouraged.

Background: the TF 2.0 versions of tf.summary.scalar(), image(), etc now require a `step` argument at the callsite. This is because the new ops emit summaries immediately to the default file writer, so there's no subsequent `add_summary(merged_summaries, step=step)` call at which the step can be provided. This is inconvenient when working with nested code in which summaries are written from parts of the code that don't have easy access to a step variable, or when working with frameworks where the logical "step" isn't easily accessible to user code.

Note that there is a provisional `tf.summary.experimental.set_step()` implementation that can be used to set a default step value or variable for all summary writing calls from that thread, but that may be changed or removed at any time because it's unclear if that's the right API for the long term.

**Comment by foxik** [comment]
Have you considered using a ContextManager for providing the default step to be used? Something very similar to `SummaryWriter.as_default`, in the lines of
```python
  with tf.summary.recording_step(step):
    ...
    tf.summary.scalar(...)
    ...
```
That seems useful enough (allowing client code to either reuse a "current" step, or provide an overload), and similar to `tf.summary.record_if` and the `as_default()`, which the users are already accustomed to.

**Comment by nfelt** [comment]
@foxik Thanks for the thoughtful input.  I did consider a context manager but in general the context managers make the API a bit cumbersome (since they force extra indentation and can't really be activated/deactivated flexibly), so since this `get/set_step()` API is very much provisional I opted to keep it simple.  (Note that's possible to implement a context manager API based on the get/set API, but hackier to go the other direction.)

I'm actually planning on rethinking how the step management works a little more broadly, and making steps associated more directly with writers rather than a separate concept.  Do you have specific use cases for setting the step in just a local context, but not via passing it explicitly to e.g. `tf.summary.scalar(..., step=mystep)`?

**Comment by foxik** [comment]
Thanks for the quick answer. Yes, I understand that implementing a context manager is more tricky, considering that also tracing should work nicely (currently I would assume the `{get,set}_step` is converted to reading/writing `global_step`).

For setting steps in local contexts, I sometimes report results in `optimizer.iterations` for training, but `epochs` for dev/val/test. If the metrics is computed by a function, it can be given the `step` and it will work nicely for both optimizer iterations and epochs, but explicitly threading the `step` arguments seems cumbersome (manually passing steps around seems similar to manually passing summary writers, which we do not require the user to do).

But the connection of step and summary writer seems like a good idea. I could imagine the `SummaryWriter.as_default(step=None)` method which can be given a default `step` value. In this case whoever wants to set default number of steps need to know the summary writer to which it should be applied, but it does not seem like a big restriction (and they can get the default SummaryWriter anyway).

**Comment by janosh** [comment]
Any progress on this?

> Background: the TF 2.0 versions of tf.summary.scalar(), image(), etc now require a step argument at the callsite. (...) This is inconvenient when working with nested code in which summaries are written from parts of the code that don't have easy access to a step variable, ...

I'm in that exact situation and agree that `SummaryWriter.as_default(step=None)` seems like a good solution here.

**Comment by foxik** [comment]
This is still the case in TF 2.1.

**Comment by janosh** [comment]
What's still the case? You mean this isn't implemented yet?

**Comment by foxik** [comment]
Yes, I meant nothing changed in TF 2.1 compared to TF 2.0 (as far as I know).

Personally I would really like an optional `step` argument to `SummaryWriter.as_default` and `set_as_default`. It would be also simple to implement. I will consider sending a pull request.

**Comment by foxik** [comment]
I created a PR with the `step` argument of `SummaryWriter.as_default`. For symmetry, I added it also to `SummaryWriter.set_as_default`.

**Comment by foxik** [comment]
With the #36839 merged, I believe this can be closed.

Cheers.

**Comment by sanatmpa1** [comment]
@nfelt,

Can you take a look at the above comment by @foxik  and confirm if we are good to close this issue? Thanks!

**Comment by nfelt** [comment]
It's true that #36839 landed - thanks @foxik for the contribution - but there's still a couple points that seem unresolved to me:

- The considerations discussed on the PR, i.e. https://github.com/tensorflow/tensorflow/pull/36839#issuecomment-652204270 - right now the added APIs still change the global step, but making the step truly per-writer would be IMO a better-encapsulated approach
- We should resolve the status of the experimental get/set step APIs; either deprecating them in favor of the new ones, or promoting them to non-experimental and just having the writer-associated and non-writer-associated ways to set the step. This is somewhat dependent on the bullet above.

So I'd suggest leaving the issue open for those two follow-ups.

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
